core/base/matrix_data.hpp

```diff
@@ -300,6 +320,69 @@ struct matrix_data {
         return res;
     }
 
+    /**
+     * Initializes a random dense matrix with a specific condition number.
+     *
+     * The matrix is generated by applying a series of random Hausholder
+     * reflectors to a diagonal matrix with diagonal entries uniformly
+     * distributed between `sqrt(condition_number)` and
+     * `1/sqrt(condition_number)`.
+     *
+     * @param size  number of rows and columns of the matrix
+     * @param condition_number  condition number of the matrix
+     * @param dist  random distribution used to generate reflectors
+     * @param engine  random engine used to generate reflectors
+     * @param num_reflectors  number of reflectors to apply from each side
+     */
+    template <typename RandomDistribution, typename RandomEngine>
+    static matrix_data cond(size_type size,
+                            remove_complex<ValueType> condition_number,
+                            RandomDistribution &&dist, RandomEngine &&engine,
+                            size_type num_reflectors)
+    {
+        using range = range<accessor::row_major<ValueType, 2>>;
+        std::vector<ValueType> mtx_data(size * size, zero<ValueType>());
+        std::vector<ValueType> ref_data(size);
+        std::vector<ValueType> work(size);
+        range matrix(mtx_data.data(), size, size, size);
+        range reflector(ref_data.data(), size, 1u, 1u);
+
+        initialize_diag_with_cond(condition_number, matrix);
+        for (size_type i = 0; i < num_reflectors; ++i) {
+            generate_random_reflector(dist, engine, reflector);
```

Comment:
Not sure I get it. Wouldn't I basically need to wrap the entire function into something and then just call that? I don't see there's a component of this that can be replaced with something else - the sizes of the arrays and the data assume it's a Hausholder reflector, the number of refectors and the size of the for loop also assume we're using Housholder reflectors.

Also not sure how I would use GS here - that thing gets you an orthogonal base. We're not orthogonalizing anything here, just scrambling up a diagonal matrix without modifying its singular values. Orthogonal transformations preserve the singular values, and Hausholder reflectors are the easiest to generate. We could alternatively use Givens rotations but I don't see a reason why would we want to do that. Even if we did - the number of rotations would have to be far greater than the reflectors (and `num_reflectors` does not make sense in that context), and the data storage would have to be completely different than for Hausholder.

Also, the functions wouldn't have `reflector` but `rotation` in the name, so that would keep them apart.

I wouldn't overdesign this right now, as I don't see any reason that we would need multiple algorithms to generate random matrices in the future.

Reply:
Okay. Sounds good.